In [51]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [6]:
raw_data = pd.read_csv('2.02. Binary predictors.csv')
raw_data.head()

SAT Admitted  Gender
0  1363       No    Male
1  1792      Yes  Female
2  1954      Yes  Female
3  1653       No    Male
4  1593       No    Male

In [8]:
data = raw_data.copy()
data['Admitted'] = data['Admitted'].map({'Yes':1,'No':0})
data['Gender'] = data['Gender'].map({'Female':1,'Male':0})
data.head()

SAT  Admitted  Gender
0  1363         0       0
1  1792         1       1
2  1954         1       1
3  1653         0       0
4  1593         0       0

In [15]:
x1 = data[['SAT','Gender']]
y = data['Admitted']

In [16]:
x = sm.add_constant(x1)
log_result = sm.Logit(y,x).fit()
log_result.summary()

Optimization terminated successfully.
         Current function value: 0.120117
         Iterations 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      165
Method:                           MLE   Df Model:                            2
Date:                Sat, 25 Jan 2020   Pseudo R-squ.:                  0.8249
Time:                        22:04:46   Log-Likelihood:                -20.180
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.118e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -68.3489     16.454     -4.154      0.000    -100.598     -36.100
SAT            0.0406      0.010      4.129      0.000       0.021       0.060
Gender         1.9449      0.846      2.299      0.022       0.287       3.603
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

In [17]:
np.exp(1.9449)

6.992932526814459

#### With the information given, though lacking context, we can say that given the same SAT score, a female is 6.99 times more likely to get admitted

### Testing the Accuracy of the Prediction

In [29]:
np.set_printoptions(formatter = {'all': lambda x: '{0:.2f}'.format(x)}) # The set_printoptions method allows us to
#format the output
log_result.predict() #We compare the results of our prediction with the actual values from the dataset

array([0.00, 1.00, 1.00, 0.23, 0.02, 0.99, 1.00, 1.00, 1.00, 0.01, 1.00,
       1.00, 0.76, 0.00, 0.60, 1.00, 0.11, 0.12, 0.51, 1.00, 1.00, 1.00,
       0.00, 0.01, 0.97, 1.00, 0.48, 0.99, 1.00, 0.99, 0.00, 0.83, 0.25,
       1.00, 1.00, 1.00, 0.31, 1.00, 0.23, 0.00, 0.02, 0.45, 1.00, 0.00,
       0.99, 0.00, 0.99, 0.00, 0.00, 0.01, 0.00, 1.00, 0.92, 0.02, 1.00,
       0.00, 0.37, 0.98, 0.12, 1.00, 0.00, 0.78, 1.00, 1.00, 0.98, 0.00,
       0.00, 0.00, 1.00, 0.00, 0.78, 0.12, 0.00, 0.99, 1.00, 1.00, 0.00,
       0.30, 1.00, 1.00, 0.00, 1.00, 1.00, 0.85, 1.00, 1.00, 0.00, 1.00,
       1.00, 0.89, 0.83, 0.00, 0.98, 0.97, 0.00, 1.00, 1.00, 0.03, 0.99,
       0.96, 1.00, 0.00, 1.00, 0.01, 0.01, 1.00, 1.00, 1.00, 0.00, 0.00,
       0.02, 0.33, 0.00, 1.00, 0.09, 0.00, 0.97, 0.00, 0.75, 1.00, 1.00,
       0.01, 0.01, 0.00, 1.00, 0.00, 0.99, 0.57, 0.54, 0.87, 0.83, 0.00,
       1.00, 0.00, 0.00, 0.00, 1.00, 0.04, 0.00, 0.01, 1.00, 0.99, 0.52,
       1.00, 1.00, 0.05, 0.00, 0.00, 0.00, 0.68, 1.

In [30]:
np.array(data['Admitted']) #The actual values of 'Admitted' from the dataset

array([0.00, 1.00, 1.00, 0.00, 0.00, 1.00, 1.00, 1.00, 1.00, 0.00, 1.00,
       1.00, 1.00, 0.00, 0.00, 1.00, 0.00, 0.00, 1.00, 1.00, 1.00, 1.00,
       0.00, 0.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 0.00, 1.00, 0.00,
       1.00, 1.00, 1.00, 0.00, 1.00, 0.00, 0.00, 0.00, 1.00, 1.00, 0.00,
       1.00, 0.00, 1.00, 0.00, 0.00, 0.00, 0.00, 1.00, 0.00, 0.00, 1.00,
       0.00, 0.00, 1.00, 0.00, 1.00, 0.00, 1.00, 1.00, 1.00, 1.00, 0.00,
       0.00, 0.00, 1.00, 0.00, 1.00, 1.00, 0.00, 1.00, 1.00, 1.00, 0.00,
       1.00, 1.00, 1.00, 0.00, 1.00, 1.00, 0.00, 1.00, 1.00, 0.00, 1.00,
       1.00, 1.00, 0.00, 0.00, 1.00, 1.00, 0.00, 1.00, 1.00, 0.00, 1.00,
       1.00, 1.00, 0.00, 1.00, 0.00, 0.00, 1.00, 1.00, 1.00, 0.00, 0.00,
       0.00, 0.00, 0.00, 1.00, 0.00, 0.00, 1.00, 0.00, 1.00, 1.00, 1.00,
       0.00, 0.00, 0.00, 1.00, 0.00, 1.00, 0.00, 1.00, 1.00, 1.00, 0.00,
       1.00, 0.00, 0.00, 0.00, 1.00, 0.00, 0.00, 0.00, 1.00, 1.00, 1.00,
       1.00, 1.00, 0.00, 0.00, 0.00, 0.00, 1.00, 1.

In [31]:
log_result.pred_table() # The pred_table method allows us to compare our predictions with the actual values
# This matrix is known as a confusion matrix

array([[69.00, 5.00],
       [4.00, 90.00]])

In [38]:
# The confusion matrix informs us of how well our model performed
# We Put it in a summary table
compare = pd.DataFrame(data = log_result.pred_table(), columns = [['Predicted 0', 'Predicted 1']])
compare = compare.rename(index = {0 :"Actual 0", 1 :"Actual 1"})
compare

Predicted 0 Predicted 1
Actual 0        69.0         5.0
Actual 1         4.0        90.0

In [42]:
acc_array = np.array(compare)
acc_array

array([[69.00, 5.00],
       [4.00, 90.00]])

In [50]:
# To calculate the accuracy of the model
model_accuracy = (acc_array[0,0]+acc_array[1,1])/(acc_array).sum()
model_accuracy *= 100
model_accuracy

94.64285714285714

### Testing the Model

In [72]:
# A set of observations have been put aside for testing. These observations were not included in the dataset used to 
# train the model
test_data = pd.read_csv('2.03. Test dataset.csv')
test_data
raw_test_data = test_data.copy()

In [73]:
x

const   SAT  Gender
0      1.0  1363       0
1      1.0  1792       1
2      1.0  1954       1
3      1.0  1653       0
4      1.0  1593       0
..     ...   ...     ...
163    1.0  1722       1
164    1.0  1750       0
165    1.0  1555       0
166    1.0  1524       0
167    1.0  1461       0

[168 rows x 3 columns]

In [74]:
# To run a prediction on the test data, we need to put them in the same format as the that of the data passed into
# the model for prediction

In [75]:
test_data = test_data.drop(['Admitted'], axis = 1)
test_data = sm.add_constant(test_data)
test_data

/usr/local/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


const   SAT  Gender
0     1.0  1323    Male
1     1.0  1725  Female
2     1.0  1762  Female
3     1.0  1777    Male
4     1.0  1665    Male
5     1.0  1556  Female
6     1.0  1731  Female
7     1.0  1809  Female
8     1.0  1930  Female
9     1.0  1708    Male
10    1.0  1970  Female
11    1.0  2039  Female
12    1.0  1681  Female
13    1.0  1444    Male
14    1.0  1726    Male
15    1.0  1876  Female
16    1.0  1485    Male
17    1.0  1500  Female
18    1.0  1900    Male

In [76]:
test_data['Gender'] = test_data['Gender'].map({'Male': 0, 'Female': 1})
test_data.head()

const   SAT  Gender
0    1.0  1323       0
1    1.0  1725       1
2    1.0  1762       1
3    1.0  1777       0
4    1.0  1665       0

In [80]:
raw_test_data['Admitted'] = raw_test_data['Admitted'].map({'Yes':1,'No':0})

In [81]:
np.array(raw_test_data['Admitted'])

array([0.00, 1.00, 1.00, 1.00, 0.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00,
       1.00, 1.00, 0.00, 0.00, 1.00, 0.00, 0.00, 1.00])

In [85]:
# Since statsmodels pred_table() method doesn't accommodate for test data, we create a function that returns the 
# confusion matrix and accuracy

def confusion_matrix(data, actual_values, model):
    pred_values = model.predict(data)
    bins = np.array([0,0.5,1]) # Create a bin to assign 0 to values less than 0.5 and 1 to values higher
    cm = np.histogram2d(actual_values, pred_values, bins)[0]
    accuracy = (cm[0,0]+cm[1,1])/cm.sum()
    return cm, accuracy

In [86]:
cm_table = confusion_matrix(test_data, raw_test_data['Admitted'], log_result)
cm_table

(array([[5.00, 1.00],
        [1.00, 12.00]]), 0.8947368421052632)

In [87]:
# Representing the confusion matrix in a table

sum_table = pd.DataFrame(data = cm_table[0], columns = [['Predicted 0','Predicted 1']])
sum_table = sum_table.rename(index = {0:'Actual 0', 1:'Actual 1'})
sum_table

Predicted 0 Predicted 1
Actual 0         5.0         1.0
Actual 1         1.0        12.0

#### As we can see, the model is 89% accurate as the model was confused 2 times in 19 predictions

In [89]:
# The misclassification rate is 
print("Misclassification rate :",str(1- cm_table[1]))

Misclassification rate : 0.10526315789473684


In [ ]:
# The training data is more accurate compared to the test data accuracy. This is a